pkg/syncer/upsync/upsync_reconcile.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			return false, err
+		}
+
+		// - finally update the main content again to set the resource version annotation to the value of the downstream resource version.
+		preparedResource.SetAnnotations(addResourceVersionAnnotation(resourceVersionDownstream, preparedResource.GetAnnotations()))
+		preparedResource.SetResourceVersion(updatedResource.GetResourceVersion())
+		_, err = c.upstreamClient.Resource(gvr).Cluster(upstreamClusterName.Path()).Namespace(upstreamNamespace).Update(ctx, preparedResource, metav1.UpdateOptions{})
+		// If the downstream resource is marked for deletion, let's requeue it to manage the deletion timestamp
+		return markedForDeletionDownstream, err
+	}
+
+	if downstreamResource.GetDeletionTimestamp() != nil {
+		return false, c.removeUpstreamResource(ctx, gvr, upstreamClusterName, upstreamNamespace, upstreamName, false)
+	}
+
+	return false, nil
+}
+
+func (c *controller) prepareResourceForUpstream(ctx context.Context, gvr schema.GroupVersionResource, upstreamNS string, upstreamLogicalCluster logicalcluster.Name, downstreamObj *unstructured.Unstructured) *unstructured.Unstructured {
+	// Make a deepcopy
+	resourceToUpsync := downstreamObj.DeepCopy()
+	annotations := resourceToUpsync.GetAnnotations()
+	if annotations != nil {
+		delete(annotations, shared.NamespaceLocatorAnnotation)
+		resourceToUpsync.SetAnnotations(annotations)
+	}
+	labels := resourceToUpsync.GetLabels()
+	if labels != nil {
+		delete(labels, workloadv1alpha1.InternalDownstreamClusterLabel)
+		resourceToUpsync.SetLabels(labels)
+	}
+	resourceToUpsync.SetNamespace(upstreamNS)
+	resourceToUpsync.SetUID("")
+	resourceToUpsync.SetResourceVersion("")
+	resourceToUpsync.SetManagedFields(nil)
+	resourceToUpsync.SetDeletionTimestamp(nil)
+	resourceToUpsync.SetDeletionGracePeriodSeconds(nil)
+	resourceToUpsync.SetOwnerReferences(nil)
+	resourceToUpsync.SetFinalizers([]string{shared.SyncerFinalizerNamePrefix + c.syncTargetKey})
+
+	return resourceToUpsync
+}
+
+func (c *controller) removeUpstreamResource(ctx context.Context, gvr schema.GroupVersionResource, clusterName logicalcluster.Name, namespace, name string, force bool) error {
+	if force {
+		existingResource, err := c.upstreamClient.Resource(gvr).Cluster(clusterName.Path()).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
+		if k8serror.IsNotFound(err) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+		if len(existingResource.GetFinalizers()) > 0 {
+			existingResource.SetFinalizers(nil)
+			if _, err := c.upstreamClient.Resource(gvr).Cluster(clusterName.Path()).Namespace(namespace).Update(ctx, existingResource, metav1.UpdateOptions{}); k8serror.IsNotFound(err) {
+				return nil
+			} else if err != nil {
+				return err
+			}
+		}
+	}
+	if err := c.upstreamClient.Resource(gvr).Cluster(clusterName.Path()).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{}); k8serror.IsNotFound(err) {
+		return nil
+	} else {
+		return err
+	}
```

Comment:
```suggestion
return c.upstreamClient.Resource(gvr).Cluster(clusterName.Path()).Namespace(namespace).Delete(ctx, name, metav1.DeleteOptions{})
```

Reply:
and move IsNotFound outside where this is called

Reply:
Fixed in commit fffff6927c99c86702c7709881a77057331950cd